lib/src/archive_descriptor.dart

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:io';
+
+import 'package:archive/archive.dart';
+import 'package:async/async.dart';
+import 'package:meta/meta.dart';
+import 'package:path/path.dart' as p;
+import 'package:test/test.dart';
+
+import 'descriptor.dart';
+import 'directory_descriptor.dart';
+import 'file_descriptor.dart';
+import 'sandbox.dart';
+import 'utils.dart';
+
+/// A [Descriptor] describing files in a Tar or Zip archive.
+///
+/// The format is determined by the descriptor's file extension.
+@sealed
+class ArchiveDescriptor extends Descriptor implements FileDescriptor {
+  /// Descriptors for entries in this archive.
+  final List<Descriptor> contents;
+
+  /// Returns a `package:archive` [Archive] object that contains the contents of
+  /// this file.
+  Future<Archive> get archive async {
+    var archive = Archive();
+    (await _files(contents)).forEach(archive.addFile);
+    return archive;
+  }
+
+  File get io => File(p.join(sandbox, name));
+
+  /// Returns [ArchiveFile]s for each file in [descriptors].
+  ///
+  /// If [parent] is passed, it's used as the parent directory for filenames.
+  Future<Iterable<ArchiveFile>> _files(Iterable<Descriptor> descriptors,
+      [String parent]) async {
+    return (await waitAndReportErrors(descriptors.map((descriptor) async {
+      var fullName =
+          parent == null ? descriptor.name : "$parent/${descriptor.name}";
+
+      if (descriptor is FileDescriptor) {
+        var bytes = await collectBytes(descriptor.readAsBytes());
+        return [
+          ArchiveFile(fullName, bytes.length, bytes)
+            // Setting the mode and mod time are necessary to work around
+            // brendan-duncan/archive#76.
+            ..mode = 428
+            ..lastModTime = DateTime.now().millisecondsSinceEpoch ~/ 1000
+        ];
+      } else if (descriptor is DirectoryDescriptor) {
+        return await _files(descriptor.contents, fullName);
+      } else {
+        throw UnsupportedError(
+            "An archive can only be created from FileDescriptors and "
+            "DirectoryDescriptors.");
+      }
+    })))
+        .expand((files) => files);
+  }
+
+  ArchiveDescriptor(String name, Iterable<Descriptor> contents)
+      : contents = List.unmodifiable(contents),
+        super(name);
+
+  Future create([String parent]) async {
```

Comment:
Should this be typed as `Future<void>`, for understandability?

Reply:
I was matching the superclass's signature. Probably the whole package should be converted to `Future<void>`, but I don't think this is the PR for it.